rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+
+This goes further than the convention of using Protobuf to define the interfaces of a particular plugin, where Protobufs only define the interface, whereas a TCK will assert behavior that lies beneath the interface. Combining a TCK and adherence to the interface, contracts will be fully asserted between a plugin and its extension point(s).
+
+### Up/Downgrading Plugins
+This section applies to federated development scenarios, where the operator is not necessarily the same as the plugin developers; where the release cadence of a plugin does not adhere to the release cadence of the core Spinnaker services.
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+When a service starts up, PF4J will scan a designated locally-accessible directory for plugins to install (these plugins may be downloaded at runtime or baked into the image at built-time). Once an application has started, it will notify Front50 of the plugins that are installed, and get a response dictating which plugins are to be enabled.
+
+In the case of a runtime-downloaded plugin, version to download will be dictated by Front50. Should Front50 be unavailable, the service will download the version defined by the service, or if none is defined, the latest plugin version that matches the Spinnaker service version.
+
+When a new plugin version is released, it is the responsibility of the operator to release this new version. _**TBD**: How will an operator know a new version exists?_ This new version will need to be released in lock-step with the services that it is associated with. In the future, this functionality _may_ be dynamically sourced and then re-activated. See [pf4j/pf4j-update](https://github.com/pf4j/pf4j-update)
+
+Tightly-coupled service and plugin releases is a poor developer and operator story, but an acceptable shortcoming for an initial release. By using Front50 as the source of truth, we’ll be fairly well equipped to make an acceptable runtime plugin loading story.
+
+Assuming a world where in-process plugins can be upgraded at runtime, services will need a background process to download these new plugins whenever a delta is detected in Front50, and the plugin refreshed. PF4J offers an abstraction for this reloading mechanism, but we would need to figure out how Spring integrations are handled. For example, it’s possible we may need to make the constraint that exported beans from a plugin are not refreshed on plugin up/down grade at least initially.
+
+#### Remote
+Remote plugin upgrade process is not much different than iterating a remote service, since that’s what it is.
+
+When a remote plugin is released, its internal functionality will change immediately, just as it would be expected for a remote service. When the plugin starts, it must send a registration event to Spinnaker. This registration event would contain, at least:
+
+* Plugin ID
```

Comment:
why not send the same manifest as https://github.com/spinnaker/community/pull/43/files#diff-d01483cbab3802f88083e6b1541851d2R115

Reply:
It probably would. This section hasn't been updated since I added that proto.

Reply:
Since, according to this, we only get updated plugin info from the remote endpoint only upon plugin start, there's a high likelihood that the remote may be updated and we'll not refresh info in Front50 until the next time the service starts. Is that important enough that we should we poll periodically?

Reply:
When the remote plugin starts, it will send its metadata to Spinnaker. This is a push model, where the remote plugin is responsible for telling Spinnaker what its metadata is. In this scenario, there is no likelihood of staleness assuming the plugin's push does not fail. The plugin can further decide to push this metadata to Spinnaker whenever it wants as well.

Reply:
Ah, this was not clear at all to me when I was reading that... I assumed you were talking about the in-process counterpart to the remote plugin here, not the actual remote service. I guess maybe if we added something to indicate that "Spinnaker will expose an API for the remote plugin to send this registration event" or maybe that "the TCK will provide registration functionality to plugin developers" (if we intend to do that, but I didn't think we did for remote plugins).